backend/test/mission_service_impl_test.cpp

```diff
@@ -1,55 +1,358 @@
+    {
+        _callback_saved_future = _callback_saved_promise.get_future();
+    }
+
+    /* The mocked mission module. */
+    MockMission _mission;
+
+    /* The mission service that is actually being tested here. */
+    MissionServiceImpl _mission_service;
+
+    /* StartMission returns its result through a callback, which is saved in _result_callback. */
+    dc::testing::mission_result_callback_t _result_callback;
+
+    /* The tests need to make sure that _result_callback has been set before calling it, hence the promise. */
+    std::promise<void> _callback_saved_promise;
+
+    /* The tests need to make sure that _result_callback has been set before calling it, hence the future. */
+    std::future<void> _callback_saved_future;
+};
+
+class MissionServiceImplUploadTest : public MissionServiceImplTestBase
+{
+protected:
+    /**
+     * Uploads the mission and saves the result callback together with the actual list of items
+     * that are sent to dronecore. The result callback is saved in _result_callback, and the
+     * mission items are saved in _uploaded_mission.
+     */
+    std::future<void> uploadMissionAndSaveParams(std::shared_ptr<UploadMissionRequest> request,
+                                                 std::shared_ptr<UploadMissionResponse> response);
+
+    /* Generate an UploadMissionRequest from a list of mission items. */
+    std::shared_ptr<UploadMissionRequest>
+    generateUploadRequest(const std::vector<std::shared_ptr<dc::MissionItem>> &mission_items) const;
 
-TEST_P(MissionServiceImplTest, uploadResultIsTranslatedCorrectly)
+    /* Translate a CameraAction into an RPCCameraAction */
+    RPCCameraAction translateCameraAction(const CameraAction camera_action) const;
+
+    /**
+     * Upload a list of items through gRPC, catch the list that is actually sent by
+     * the backend, and verify that it has been sent correctly over gRPC.
+     */
+    void checkItemsAreUploadedCorrectly(std::vector<std::shared_ptr<dc::MissionItem>> &mission_items);
+
+    /* Generate a list of one mission item. */
+    std::vector<std::shared_ptr<dc::MissionItem>> generateListOfOneItem();
+
+    /* Generate a list of multiple mission items. */
+    std::vector<std::shared_ptr<dc::MissionItem>> generateListOfMultipleItems();
+
+    /* Captures the actual mission sent to dronecore by the backend. */
+    std::vector<std::shared_ptr<dc::MissionItem>> _uploaded_mission;
+};
+
+INSTANTIATE_TEST_CASE_P(MissionResultCorrespondences,
+                        MissionServiceImplUploadTest,
+                        ::testing::ValuesIn(generateInputPairs()));
+
+ACTION_P3(SaveUploadParams, mission, callback, callback_saved_promise)
+{
+    *mission = arg0;
+    *callback = arg1;
+    callback_saved_promise->set_value();
+}
+
+TEST_F(MissionServiceImplUploadTest, doesNotFailWhenArgsAreNull)
 {
-    // TODO
+    auto upload_handle = uploadMissionAndSaveParams(nullptr, nullptr);
+    _result_callback(dc::Mission::Result::UNKNOWN);
```

Comment:
Why would the result be `UNKNOWN`? I think the result should only be unknown if something in the enum translation went wrong. If it is an actual error, it should be `ERROR` or some specific error code, right?

Or is it that the error is unknown because with `nullptr` the callback has not actually been called?

Reply:
It is just that I need to synchronize the call in the service (see [here](https://github.com/dronecore/DroneCore/blob/350-mission-backend/backend/src/plugins/mission/mission_service_impl.h#L30) and [here](https://github.com/dronecore/DroneCore/blob/350-mission-backend/backend/src/plugins/mission/mission_service_impl.h#L102)). So the upload call is blocked until the callback is actually called (this assumes that the callback is always called, eventually).

The value of it in this unit test has not importance. It just needs to trigger. Maybe I should make it a variable called `ARBITRARY_RESULT`, to make this clear!